torch/_inductor/codegen/common.py

```diff
@@ -1069,6 +1073,34 @@ def load(name: str, index: sympy.Expr):
                     return store_cache[name]
                 return self.load(name, index)
 
+            @staticmethod
+            def check_bounds(var, expr, size):
+                ranges = self.get_ranges()  # type: ignore[attr-defined]
+
+                should_compute_bounds = True
+                if isinstance(size, int) or size.is_number:
+                    if "Where" in str(expr):
+                        expr_ = expr.subs({expr.args[-1]: 0})
+                        bounds = bound_sympy(expr_, ranges)
+                        # bounds must be within the range [-size, size) to be valid
+                        if bounds.lower >= -size and bounds.upper < size:
+                            should_compute_bounds = False
+                    else:
+                        bounds = bound_sympy(expr, ranges)
+                        # bounds in [0, size)
+                        if bounds.lower >= 0 and bounds.upper < size:
+                            should_compute_bounds = False
+
+                if should_compute_bounds:
+                    bounds = bound_sympy(expr, ranges)
+                    index = self.index_to_str(expr)  # type: ignore[attr-defined]
+                    new_var = self.cse.generate(self.compute, index, bounds=bounds)
+                    new_var.update_on_args("index_wrap", (var,), {})
+                    # set the bounds on new_var as "update_on_args" may change
+                    # the bounds to "var.bounds"
+                    new_var.bounds = bounds
+                    return CSEProxy.indirect_indexing(new_var, size)
```

Comment:
There are several issues in this function at first sight:
- You are recomputing the bounds of `expr` for no good reason
- You are calling this function with an `expr` that you know that has a where, but then you remove it (in a far from efficient way).
- You are then forwarding the expression with `Where` to `indirect_indexing`, when that function will itself wrap things with a ternary operator if necessary

Please try to find a way to streamline all the logic here. Start by factorising out the logic that wraps an expression into its own function... like `check_bounds`...
Then, that `check_bounds` function can the index wrapped properly (let that index be a sympy index or an indirect index, it will need to know how to handle both)
-

Reply:
> You are recomputing the bounds of expr for no good reason

This is for the cases I can prove the runtime check is not needed.

> You are calling this function with an expr that you know that has a where, but then you remove it (in a far from efficient way).

There are cases that have no `Where`, as SymPy was able to prove the expression was greater than 0. Also, I've changed the way I was using to detect and replace the `Where`.